rmf_fleet_adapter/src/lift_supervisor/Node.cpp

```diff
@@ -53,14 +53,49 @@ Node::Node()
 //==============================================================================
 void Node::_adapter_lift_request_update(LiftRequest::UniquePtr msg)
 {
-  // For now we will simply forward the request along
+  if (_active_sessions.find(msg->lift_name) == _active_sessions.end())
+    _active_sessions[msg->lift_name] = nullptr;
+  
+  auto& curr_request = _active_sessions[msg->lift_name];
+  if (curr_request)
+  {
+    if (curr_request->session_id == msg->session_id)
+    {
+      if (msg->request_type != LiftRequest::REQUEST_END_SESSION)
+        curr_request = std::move(msg);
+      else
+      {
+        _lift_request_pub->publish(*msg);
+        curr_request = nullptr;
+      }
+    }
+  }
+  else
+  {
+    if (msg->request_type != LiftRequest::REQUEST_END_SESSION)
+    {
+      curr_request = std::move(msg);
+    }
+  }
+
   // TODO(MXG): Make this more intelligent by scheduling the lift
-  _lift_request_pub->publish(*msg);
 }
 
 //==============================================================================
-void Node::_lift_state_update(LiftState::UniquePtr /*msg*/)
+void Node::_lift_state_update(LiftState::UniquePtr msg)
 {
+  if (_active_sessions.find(msg->lift_name) == _active_sessions.end())
+    _active_sessions[msg->lift_name] = nullptr;
+
+  auto& lift_request = _active_sessions[msg->lift_name];
```

Comment:
Similar to the earlier comment, this is doing 3 map lookups when you could reduce it to just one.

```
auto& lift_request = _active_sessions.insert(std::make_pair<std::string, LiftRequest::UniquePtr>(msg->lift_name, nullptr)).first->second;
```